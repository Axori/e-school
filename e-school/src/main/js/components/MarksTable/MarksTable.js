import React, {useEffect, useState} from 'react';
import {cloneDeep} from "lodash";
import Loader from "../Loader/Loader";

const MarksTable = ({loading, studentMarks, onSave, isStudent}) => {
    const [studentMarksCopy, setStudentMarksCopy] = useState();
    const [hasChanges, setHasChanges] = useState(false);
    const numberOfMarkColumns = 10;

    useEffect(() => {
        if (studentMarks) {
            const newStudentMarks = cloneDeep(studentMarks);
            Object.keys(newStudentMarks).map((key) => {
                const {marks} = newStudentMarks[key];
                const lengthDifference = numberOfMarkColumns - marks.length;
                const numberOfExtraInputs = lengthDifference >= 0 ? lengthDifference : 0;

                const newMarks = [...marks, ...(new Array(numberOfExtraInputs).fill(0))];
                newStudentMarks[key].marks = newMarks;
            });

            setStudentMarksCopy(newStudentMarks);
            setHasChanges(false);
        }
    }, [studentMarks]);

    const handleOnChange = (id, index) => (e) => {
        if (e.target.checkValidity()) {
            const newValue = parseInt(e.target.value);
            const newStudentMarks = cloneDeep(studentMarksCopy);
            newStudentMarks[id].marks.splice(index, 1, newValue);
            setStudentMarksCopy(newStudentMarks);
            setHasChanges(true);
        }
    }

    const handleOnSave = () => {
        onSave(Object.entries(studentMarksCopy).reduce((prev, [id, studentMark]) => ({
            ...prev,
            [id]: {
                ...studentMark,
                marks: studentMark.marks.filter(el => el > 0)
            }
        }), {}))
    }
    console.log("loading, marks, marksCopy", studentMarks, studentMarksCopy)
    return <div className="MarksTable d-flex justify-content-center">
        {!loading && studentMarksCopy ? <div className="MarksTable__container">
                <table id="MarksTable" className="table">
                    <thead>
                    <tr>
                        <th scope="col">Name</th>
                        <th scope="col" colSpan={numberOfMarkColumns}>Marks</th>
                    </tr>
                    </thead>
                    <tbody>
                    {Object.entries(studentMarksCopy).map(([id, {name, marks}]) =>
                        <tr key={id}>
                            <th scope="row">{name}</th>
                            {marks.map((mark, index) =>
                                <td className="MarksTable__input" key={`${name}-mark-${index}`}>
                                    {!isStudent ?
                                        <input className="form-control" type="number" min={0} max={6}
                                               value={mark}
                                               onChange={handleOnChange(id, index)}/> : mark}
                                </td>
                            )}
                        </tr>
                    )}
                    </tbody>
                </table>
                {!isStudent && hasChanges && <button className="btn btn-success" onClick={handleOnSave}>Save</button>}
            </div>
            : <Loader/>}
    </div>
}

export default MarksTable;