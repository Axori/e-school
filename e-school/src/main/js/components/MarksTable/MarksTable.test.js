import React from 'react';
import MarksTable from "./MarksTable";
import {act, create} from 'react-test-renderer';

describe("MarksTable", () => {
    const studentMarks = {
        15: {
            name: "Student 15",
            marks: [
                1, 2
            ]
        },
        20: {
            name: "Student 20",
            marks: [
                3, 4
            ]
        },

    };
    const defaultProps = {
        loading: false,
        studentMarks,
        onSave: jest.fn(),
        isStudent: false
    }

    const getComponent = (props = {}) => create(<MarksTable {...defaultProps} {...props} />)

    it("should render MarksTable with students inside", () => {
        let component;
        act(() => {
            component = getComponent();
        })

        expect(component.toJSON()).toMatchSnapshot();
    })

    it("should render Loading", () => {
        const component = getComponent({loading: true});

        expect(component.toJSON()).toMatchSnapshot();
    })
})