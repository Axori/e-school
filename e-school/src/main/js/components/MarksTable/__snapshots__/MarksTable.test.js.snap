// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarksTable should render Loading 1`] = `
<div
  className="MarksTable d-flex justify-content-center"
>
  <div
    className="spinner-border"
    role="status"
  >
    <span
      className="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;

exports[`MarksTable should render MarksTable with students inside 1`] = `
<div
  className="MarksTable d-flex justify-content-center"
>
  <div
    className="MarksTable__container"
  >
    <table
      className="table"
      id="MarksTable"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Name
          </th>
          <th
            colSpan={10}
            scope="col"
          >
            Marks
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            Student 15
          </th>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={1}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={2}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            Student 20
          </th>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={3}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={4}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
          <td
            className="MarksTable__input"
          >
            <input
              className="form-control"
              max={6}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
